newm-server/src/main/kotlin/io/newm/server/features/marketplace/repo/MarketplaceRepositoryImpl.kt

```diff
@@ -68,6 +100,118 @@ internal class MarketplaceRepositoryImpl(
         }
     }
 
+    override suspend fun generateOrderAmount(request: OrderAmountRequest): OrderAmountResponse {
+        log.debug { "generateOrderAmount: $request" }
+
+        val sale = transaction { MarketplaceSaleEntity[request.saleId] }
+        require(request.bundleQuantity <= sale.availableBundleQuantity) { "Not enough bundles available for sale" }
+
+        val minIncentiveAmount = configRepository.getLong(CONFIG_KEY_MARKETPLACE_MIN_INCENTIVE_AMOUNT)
+        val incentiveAmount =
+            request.incentiveAmount?.also {
+                require(it < minIncentiveAmount) { "Not enough incentive" }
+            } ?: minIncentiveAmount
+
+        val nativeAssets =
+            listOf(
+                nativeAsset {
+                    policy = sale.costPolicyId
+                    name = sale.costAssetName
+                    amount = (sale.costAmount.toBigInteger() * request.bundleQuantity.toBigInteger()).toString()
+                },
+                nativeAsset {
+                    policy = incentivePolicyId
+                    name = incentiveAssetName
+                    amount = (incentiveAmount.toBigInteger() * BigInteger.TWO).toString()
```

Comment:
Why are we multiplying by 2?

Reply:
Quinn explained to me that we have 2 actions: purchase & remove. So it requires 2 incentive units.